### DISCUSSION TITLE
Any way to pass custom error types from action?

### DISCUSSION THREAD

**Discussion by mattoni** [remix-run/remix#10354] [Q&A]
With the new singleFetch, is there any way to serialize and properly send a custom error class?

For example, I have this class:

```
export class ValidationError extends Error {
	public readonly fieldErrors: Record<string, string[]> = {};
	public readonly status = 422;

	constructor(fieldErrors: Record<string, string[]>, globalError?: string) {
		super(globalError);
		if (fieldErrors) {
			this.fieldErrors = fieldErrors;
		}
		this.name = "ValidationError";
	}

	public static fromZodError<T>(error: ZodError<T>) {
		return new ValidationError(
			error.flatten().fieldErrors as Record<string, string[]>,
		);
	}
}
```

Now, regarding the tidbit that any custom error class will work as long as it has a globally accessible constructor client side, I  added this to my entry.client.tsx:

```
if (typeof window !== "undefined") {
	(window as any).ValidationError = ValidationError;
}
```

And I'm getting a ValidationError type in the browser! However, fieldErrors is unpopulated. Is there a way to populate the custom fields?

**Comment by akomm** [comment]
Why don't you just return the data? Just give the client the data it needs. For success/failure case, just give the appropriate tagged result.

class instances from server simply get serialized as PJO. If you want it to be hydrated into a class, you need to write the logic on the client.
